### ISSUE TITLE
wireguard: Stopping service has no effect

### ISSUE THREAD

**Issue by nh2** [NixOS/nixpkgs#123203]
After `systemctl stop wireguard-wg0.service` I still get

```
# netstat -anup | grep 51820
udp        0      0 0.0.0.0:51820           0.0.0.0:*                           -                   
udp6       0      0 :::51820                :::*                                -
```

So Wireguard is still running in the kernel, listening on the port.

I had to do `ip link set down dev wg0` to stop it and free the port.

I believe that `systemctl stop` should stop wireguard, and free the port.

---

These are the `journalctl -f` logs of the `stop` operation:

```
# systemctl stop wireguard-wg0.service 
Warning: Stopping wireguard-wg0.service, but it can still be activated by:
  wireguard-wg0.path
```

```
May 16 06:36:47 myhost systemd[1]: Stopping WireGuard Peer - wg0 - mypeerid_here...
May 16 06:36:47 myhost systemd[1]: wireguard-wg0-peer-mypeeriid_here.service: Succeeded.
May 16 06:36:47 myhost systemd[1]: Stopped WireGuard Peer - wg0 - mypeerid_here.
May 16 06:36:47 myhost systemd[1]: Stopping WireGuard Tunnel - wg0...
May 16 06:36:47 myhost avahi-daemon[1249]: Withdrawing workstation service for wg0.
May 16 06:36:48 myhost systemd[1]: wireguard-wg0.service: Succeeded.
May 16 06:36:48 myhost systemd[1]: Stopped WireGuard Tunnel - wg0.
May 16 06:36:48 myhost systemd[1]: Starting WireGuard Tunnel - wg0...
May 16 06:36:48 myhost systemd-udevd[61430]: ethtool: autonegotiation is unset or enabled, the speed and duplex are not writable.
May 16 06:36:48 myhost systemd-udevd[61430]: Using default interface naming scheme 'v245'.
May 16 06:36:48 myhost systemd[1]: Finished WireGuard Tunnel - wg0.
May 16 06:36:48 myhost systemd[1]: Starting WireGuard Peer - wg0 - mypeerid_here...
May 16 06:36:48 myhost systemd[1]: Finished WireGuard Peer - wg0 - mypeerid_here.
```

As shown, the tunnel and peer services both immediately restart, thus effectively ignoring the user's request to `stop`.

## Technical details

On NixOS 20.09.

**Comment by nh2** [comment]
Ah, I found the reason: It's the usage of a systemd `.path` for the `privateKeyFile`, which I believe is probably wrong here.

@arianvp already commented on this `.path` unit: https://github.com/NixOS/nixpkgs/pull/60138#commitcomment-33570986

I believe the intent of PR #60138, which added it, was to track the initial creation/existence of the private key file.

However, that's not really what systemd path files are for; they are more more for things like PID files or sockets which shall exist exactly when the associated service is running; see the [docs](https://www.freedesktop.org/software/systemd/man/systemd.path.html):

> When a service unit triggered by a path unit terminates (regardless whether it exited successfully or failed), monitored paths are checked immediately again, **and the service accordingly restarted instantly**.

This means that the `.path` unit of the `privateKeyFile` immediately restarts wireguard upon shutting down.

That's probably not intended.

---

Thus, as a current workaround, the following shuts down wireguard:

```
sudo systemctl stop wireguard-wg0.path
```

---

The `privateKeyFile` is configuration, so like for other services, `restartTriggers` would probably be more appropriate than a `path` unit.

**Comment by nh2** [comment]
I made a fix in PR #123270
